File: backend/api/src/altasai/evaluation/evaluateEntities.ts
import samples from '../datasets/entity-samples.json';
import { extractEntities } from '../nlp/entityExtractor';
import { average, exactSetScore, makeResult } from './metrics';

export const evaluateEntities = () => {
  const scores = samples.map((sample) => {
    const actual = extractEntities(sample.input).map((entity) => entity.type);
    return exactSetScore(sample.expectedEntities, actual);
  });
  return makeResult('entity_extraction', average(scores), 0.75, {
    averageExactTypeCoverage: average(scores),
    cases: samples.length,
  });
};

File: backend/ml-service/app/evaluation/evaluate_entities.py
import json

from app.config import DATASET_DIR
from app.evaluation.metrics import pass_result
from app.nlp.entity_extractor import extract_entities


def evaluate() -> dict:
    samples = json.loads((DATASET_DIR / "entity_samples.json").read_text(encoding="utf-8"))
    scores = []
    for sample in samples:
        actual = {entity["type"] for entity in extract_entities(sample["text"])["entities"]}
        expected = set(sample["entities"])
        scores.append(len(actual & expected) / max(1, len(expected)))
    score = sum(scores) / max(1, len(scores))
    return pass_result("entity_extractor", score, 0.8, {"caseScores": scores})
